### PR TITLE
Add watcher for subnets

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds a new subnets watcher on the network domain. This watcher is formed by composing one watcher out of two: one for the subnet table and the second one for the subnet_association table. The second one is needed to watch underlay fan subnets.

This watcher is wired on the apiserver for its usage on the firewaller worker.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

_Note: ideally, this patch must be QA'd on a substrate supporting port opening (e.g. AWS), but we currently have an issue on model creation which is under investigation._

Bootstrap on lxd, add a model and deploy apache2:
```
juju bootstrap localhost c
juju add-model m --config logging-config="juju.worker.firewaller=DEBUG"
juju deploy apache2
```
Once the app is running, expose apache2:
```
juju expose apache2 --to-spaces alpha
```
Then create a new space (make sure you use a correct subnet in your environment) and watch for firewaller debug logs:
```
juju add-space foo 10.68.58.0/24
juju  debug-log --replay --level DEBUG --tail | grep "firewaller"     
...
controller-0: 11:52:20 DEBUG juju.worker.firewaller updating 1 units after changes in subnets
```

## Links

**Jira card:** JUJU-5927

### CODE REVIEW

File: domain/network/service/watcher.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		},
+		watcherFactory: watcherFactory,
+	}
+}
+
+// Watch returns a watcher that observes changes to subnets and their
+// association (fan underlays), filtered based on the provided list of subnets
+// to watch.
+func (s *WatchableService) WatchSubnets(ctx context.Context, subnetUUIDsToWatch set.Strings) (watcher.StringsWatcher, error) {
+	if s.watcherFactory != nil {
```
Comment (SimonRichardson) [inline domain/network/service/watcher.go:44]: Get rid of this nil check. You should always have the watcher factory.

**Review by SimonRichardson — COMMENTED** [review]
You need to add an integration test that tests your watchers. All the way from service to state and back again. See: https://github.com/juju/juju/blob/main/domain/objectstore/watcher_test.go

File: domain/network/service/watcher.go (hunk 1 above)
Comment (nvinuesa) [reply domain/network/service/watcher.go]: Done.

**Comment by nvinuesa** [comment]
@SimonRichardson I added `domain/network/watcher_test.go` to test the watcher.

**Review by SimonRichardson — APPROVED** [review]
Noice.

**Pull request merged**